request.go

```diff
@@ -0,0 +1,153 @@
+	"github.com/notaryproject/tspclient-go/internal/oid"
+)
+
+// MessageImprint contains the hash of the datum to be time-stamped.
+//
+//	MessageImprint ::= SEQUENCE {
+//	 hashAlgorithm   AlgorithmIdentifier,
+//	 hashedMessage   OCTET STRING }
+type MessageImprint struct {
+	HashAlgorithm pkix.AlgorithmIdentifier
+	HashedMessage []byte
+}
+
+// Request is a time-stamping request.
+//
+//	TimeStampReq ::= SEQUENCE {
+//	 version         INTEGER                 { v1(1) },
+//	 messageImprint  MessageImprint,
+//	 reqPolicy       TSAPolicyID              OPTIONAL,
+//	 nonce           INTEGER                  OPTIONAL,
+//	 certReq         BOOLEAN                  DEFAULT FALSE,
+//	 extensions      [0] IMPLICIT Extensions  OPTIONAL }
+type Request struct {
+	Version        int // fixed to 1 as defined in RFC 3161 2.4.1 Request Format
+	MessageImprint MessageImprint
+	ReqPolicy      asn1.ObjectIdentifier `asn1:"optional"`
+	Nonce          *big.Int              `asn1:"optional"`
+	CertReq        bool                  `asn1:"optional,default:false"`
+	Extensions     []pkix.Extension      `asn1:"optional,tag:0"`
+}
+
+// RequestOptions provides options for caller to create a new timestamp request
+type RequestOptions struct {
+	// Content is the datum to be time stamped. REQUIRED.
+	Content []byte
+
+	// HashAlgorithm is the hash algorithm to be used to hash the Content.
+	// REQUIRED and MUST be an available hash algorithm.
+	HashAlgorithm crypto.Hash
+
+	// HashAlgorithmParameters is the parameters for the HashAlgorithm.
+	// OPTIONAL.
+	HashAlgorithmParameters asn1.RawValue
+
+	// ReqPolicy specifies the TSA policy ID. OPTIONAL.
+	ReqPolicy asn1.ObjectIdentifier
+
+	// Nonce is a large random number with a high probability that the client
+	// generates it only once. The same nonce is included and validated in the
+	// response. OPTIONAL.
+	Nonce *big.Int
+
+	// CertReq determines if TSA signing certificate is included in the response.
+	// OPTIONAL.
+	CertReq bool
+
+	// Extensions is a generic way to add additional information
+	// to the request in the future. OPTIONAL.
+	Extensions []pkix.Extension
+}
+
+// NewRequest creates a timestamp request based on caller provided options.
+func NewRequest(opts RequestOptions) (*Request, error) {
+	if opts.Content == nil {
+		return nil, &MalformedRequestError{Msg: "content to be time stamped cannot be empty"}
+	}
+	hashAlg, err := oid.FromHash(opts.HashAlgorithm)
+	if err != nil {
+		return nil, &MalformedRequestError{Msg: err.Error()}
+	}
+	digest, err := hashutil.ComputeHash(opts.HashAlgorithm, opts.Content)
+	if err != nil {
+		return nil, &MalformedRequestError{Msg: err.Error()}
+	}
+	return &Request{
+		Version: 1,
+		MessageImprint: MessageImprint{
+			HashAlgorithm: pkix.AlgorithmIdentifier{
+				Algorithm:  hashAlg,
+				Parameters: opts.HashAlgorithmParameters,
+			},
+			HashedMessage: digest,
+		},
+		ReqPolicy:  opts.ReqPolicy,
+		Nonce:      opts.Nonce,
```

Comment:
IMO, nonce should not be empty  unless explicitly indicated by user, as it provides protection against replay attack.

Reply:
According to [RFC 3161 2.4.1](https://datatracker.ietf.org/doc/html/rfc3161#section-2.4.1), `nonce` is `OPTIONAL`.
However, in `notation-core-go`, we are always including a `nonce` when generating a tsq. You will see it once the `notation-core-go` PR is created.

Reply:
Yes but we can change defaults to be more secure.

Reply:
I might misunderstand your question, so what's your proposed default? If user does not provide any nonce, return an error?

Reply:
Unless user explicitly specify that they dont want nonce, we should always create nonce.